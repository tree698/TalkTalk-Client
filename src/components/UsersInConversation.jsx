import React, { useEffect, useState } from 'react';
import toast from 'react-hot-toast';
import { useLocation } from 'react-router-dom';
import { useApiContext } from '../context/ApiContext';
import { useQuery } from '@tanstack/react-query';
import Avatar from './Avatar';

export default function UsersInConversation({ onAvatarClick }) {
  const STYLE_FOR_AVATAR =
    'w-6 h-6 md:w-7 md:h-7 lg:w-8 lg:h-8 mr-1 md:mr-2 lg:mr-3 rounded-full ';

  const [length, setLength] = useState();
  const { tweetService } = useApiContext();
  const {
    state: { id },
  } = useLocation();

  const { error, data } = useQuery(
    ['usersInConversation'],
    () =>
      tweetService.getTweets('', id).then((data) => {
        const result = removeDuplicatedUser(data);
        setLength((prev) => result.length);
        return result;
      }),
    {
      enabled: !!'userInConversation',
      refetchInterval: 1000,
    }
  );

  useEffect(() => {
    error && toast.error(error);
  }, [error]);

  return (
    <section className="flex items-center">
      <span className="w-6 h-6 md:w-7 md:h-7 lg:w-8 lg:h-8 text-sm md:text-base lg:text-lg text-center mr-1 md:mr-2 lg:mr-3 border border-darkGray rounded-full">
        {length}
      </span>
      <ul>
        {data &&
          data.map((user) => (
            <button key={user.id} onClick={() => onAvatarClick(user.username)}>
              <Avatar
                photo={user.photo}
                username={user.username}
                styleForAvatar={STYLE_FOR_AVATAR}
              />
            </button>
          ))}
      </ul>
    </section>
  );
}

export function removeDuplicatedUser(array) {
  const photo = [];
  const result = [];
  array.forEach((a) => {
    if (photo.includes(a.photo)) {
      return;
    }
    photo.push(a.photo);
    result.push({ id: a.id, photo: a.photo, username: a.username });
  });
  return result;
}
