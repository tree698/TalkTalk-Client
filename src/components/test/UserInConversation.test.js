import { Route } from 'react-router-dom';
import { withApiContext, withQuery, withRouter } from '../../test/utils';
import UsersInConversation from '../UsersInConversation';
import { render, screen, waitFor } from '@testing-library/react';
import { fakeUsersData } from '../../test/usersData';
import userEvent from '@testing-library/user-event';
import { removeDuplicatedUser } from '../UsersInConversation';

describe('UserInConversation', () => {
  const fakeContext = {
    tweetService: {
      getTweets: jest.fn(),
    },
  };

  afterEach(() => fakeContext.tweetService.getTweets.mockReset());

  it('renders with data', async () => {
    const mockOnClick = jest.fn();
    fakeContext.tweetService.getTweets.mockImplementation(() => {
      const result = removeDuplicatedUser(fakeUsersData);
      return result;
    });
    const { asFragment } = render(
      withQuery(
        withApiContext(
          withRouter(
            <Route
              path="/"
              element={<UsersInConversation onAvatarClick={mockOnClick} />}
            />,
            {
              state: {
                id: 'fakeId',
              },
            }
          ),
          fakeContext
        )
      )
    );

    await waitFor(() => screen.queryByRole('button'));

    // userEvent.click(button);

    expect(asFragment()).toMatchSnapshot();
    // await waitFor(() => expect(mockOnClick).toHaveBeenCalledTimes(1));
    // await waitFor(() => expect(button).toBeInTheDocument());
  });
});
