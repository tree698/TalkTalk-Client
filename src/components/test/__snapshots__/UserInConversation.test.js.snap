// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInConversation renders with data 1`] = `
<DocumentFragment>
  <section
    class="flex items-center"
  >
    <span
      class="w-6 h-6 md:w-7 md:h-7 lg:w-8 lg:h-8 text-sm md:text-base lg:text-lg text-center mr-1 md:mr-2 lg:mr-3 border border-darkGray rounded-full"
    />
    <ul />
  </section>
</DocumentFragment>
`;
